### DISCUSSION TITLE
Fixing refraction

### DISCUSSION THREAD

**Discussion by DiegoSynth** [godotengine/godot-proposals#9580] [3D]
At the moment Godot 4 offers an in-built refraction in the materials that has a few issues:

![image](https://github.com/godotengine/godot-proposals/assets/19650031/927ea05e-24c0-46f9-ad87-b301ccfa44af)

As shown in the picture, refraction exposes a problem when used together with alpha, where objects in front of the material in question are mirrored. Another issue is the very high level of distortion directly proportional to the camera distance (if you move the camera back a few meters, the reflection goes bananas).

These issues (mainly the mirror effect) render reflection unusable (as it needs to be used together with alpha, and it's supposed to be an effect that increases realism).
Some "dither transparency" shaders can be found online (to workaround the usage of alpha) but they no longer work on Godot 4 (as far as it was tested).

What do you guys think?

**Comment by Calinou** [comment]
These artifacts are inherent to the use of screen-space refraction. To my knowledge, there isn't really any alternative that's usable in real-time rendering performance-wise. You can make certain objects not show up in the refraction by setting their depth draw mode to **Never**, but this can introduce other rendering issues.

Do you use a normal map for the refraction? What is your refraction strength value set to? For a convincing glass effect, I generally suggest using a normal map but keeping refraction strength close to 0.

> Another issue is the very high level of distortion directly proportional to the camera distance (if you move the camera back a few meters, the reflection goes bananas).

It might be possible to exclude certain pixels from the refraction based on the depth buffer (similar to what screen-space reflections do), but this sounds nontrivial and it'd have an added performance cost.

**Comment by DiegoSynth** [comment]
Hello Calinou, thanks for your answer.

I'm using the minimum (positive) value of 0.01 for Refraction scale.
I'm not applying any normal map, but a roughness map (to get some reflections on the glasses).

I understand your point; I know that with screen space it's only tricks what can be done...
I attach here 2 screenshots where I show why I was trying to use refraction (mainly because of curved glasses).
At the moment I disabled refraction, but tinted a bit the glasses, so they are not totally transparent. I would say it's ok. Maybe in the future we get raytraced refractions and all of this is solved :)

![Untitled-2](https://github.com/godotengine/godot-proposals/assets/19650031/2d89ced3-4965-4182-bccd-4ecef34457e9)

**Comment by Meccanizer** [comment]
I've run into a similar issue to this where stuff not below the mesh are being refracted using the standard Godot refractions. I've managed to figure out how to fix the refractions refracting things not underwater by using the depth buffer. If you convert the material with refractions enabled you can gain access to the refraction code.

Below is the code I used. Excuse the mess haven't cleaned it up yet.
```
//ref_ofs distorts the screen_uv for refractions
vec2 ref_ofs = SCREEN_UV- ref_normal.xy * dot(mix(texture(texture_refraction, base_uv + TIME * time),texture(texture_refraction2,base_uv+ TIME * -time),0.5), refraction_texture_channel) * refraction;

//We use ref_ofs as the UV instead of screen_uv
	float depth_tex = textureLod(depth_texture, ref_ofs, 0.0).r;

	vec4 world_pos = INV_PROJECTION_MATRIX * vec4(SCREEN_UV * 2.0 - 1.0, depth_tex2, 1.0);
	world_pos.xyz /= world_pos.w;
	float ref_amount = 1.0;

//Check if the depth buffer is greater then the models z position, if so then use the normal screen_uv otherwise use the refracted uv
	if (world_pos.z  > VERTEX.z){

		EMISSION += textureLod(screen_texture, SCREEN_UV, ROUGHNESS * 4.0).rgb * albedo.rgb * ref_amount * EXPOSURE;


	}else{

		EMISSION += textureLod(screen_texture, ref_ofs, ROUGHNESS * 4.0).rgb * albedo.rgb* ref_amount * EXPOSURE;


	}
```

Refractions without using the depth buffer. Notice the distortion on the part of the mesh not underwater
![image](https://github.com/godotengine/godot-proposals/assets/28959065/d22101a8-e0ce-4272-a7c7-b5b22dce5812)

Refractions using the depth buffer. Only meshes underwater will be refracted
![image](https://github.com/godotengine/godot-proposals/assets/28959065/ea86cfaa-20ca-4236-9741-8997571a6f12)

Only issue with this is you can't you MSAA or you will get weird artifacts around the edges of the mesh and there are some artifacts around the contact point between the water and mesh. I don't know if this will help with your usage but it will fix for the most part stuff in front of the mesh being refracted.

**Reply by DiegoSynth** [reply]
Hello Meccanizer,

Wow, this looks really good, congratulations! Thank you very much for sharing the code, that's really very kind :)

I will be trying this as soon as I can, and may come back with one or two questions (I'm relatively new to Godot, and haven't really programmed shaders in a long, long time, since XNA days), but from what I see in your code comments, it looks sensible.

**Reply by Calinou** [reply]
I opened a PR to incorporate a modified form of this suggestion: https://github.com/godotengine/godot/pull/93449

Thanks a lot for posting the code!

**Reply by DiegoSynth** [reply]
Thank you very much both of you @Meccanizer and @Calinou for making this possible in Godot!! :)
Very happy to see how the engine improves thanks to its people!
